File: algorithm/Depth First Search/dfs.c
// use Graph List
#include<stdio.h>
#define MAX 10
struct node {
  int edge;
  int visited;
  struct node *next;
};
struct node* graph[MAX];
struct node* begin[MAX];
struct node * newNode(struct node* Node,int edge)
{
  Node=(struct node*)malloc(sizeof(struct node));
  Node->edge = edge;
  Node->visited = 0;
  return Node;
}

void addNode(int vector,int edge)
{
  struct node *tail;
  tail = newNode(&tail,edge);
  if(graph[vector]==NULL) {
    tail->next=NULL;
    graph[vector]=tail;
    begin[vector]=graph[vector];
  } else {
    tail->next=graph[vector]->next;
    graph[vector]->next=tail;
  }
}

void showGraph(int i) {
  printf("index %d : ",i);
  while(graph[i] != NULL) {
    printf("%d ",graph[i]->edge);
    graph[i] = graph[i]->next;
  }
  printf("\n");
}

void setStartIndexGraph() {
  int i;
  for(i=0;i < MAX;i++) {
    graph[i] = begin[i];
  }
}

void dfs(int current) {
  if(graph[current]) {
    if(graph[current]->visited == 1) {
    return;
    }
    graph[current]->visited = 1;
    while(graph[current]) {
      dfs(graph[current]->edge);
      graph[current] = graph[current]->next;
    }
  }
  printf("%d ",current);
}

int main() {
  int i;
  for(i=0;i < MAX;i++) {
    graph[i] = NULL;
  }
  addNode(0,5);
  addNode(0,7);
  addNode(7,8);
  addNode(7,9);
  // for(i=0;i < MAX;i++) {
  //   showGraph(i);
  // }
  // setStartIndexGraph();
  dfs(0);
  printf("\n");
}

File: algorithm/Depth First Search/dfs.js
class Node {
  constructor(edge, next) {
    this.edge = edge
    this.next = next
  }
}

class Linklist {
  constructor(edge) {
    this.head = new Node(edge, null)
    this.visited = false
  } 

  add = (edge) => {
    const node = new Node(edge, null)
    let tail = this.head
    while(tail.next != null) {
      tail = tail.next
    }
    tail.next = node
  }
}

const graph = []

const createGraph = (nodes) => {
  for(let i = 0; i < nodes.length; i++) {
    if(!graph[nodes[i].vector]) {
      graph[nodes[i].vector] = new Linklist(nodes[i].edge)
    } else {
      graph[nodes[i].vector].add(nodes[i].edge)
    }
  }
}

createGraph([
  { vector: 1, edge: 4 },
  { vector: 1, edge: 2 },
  { vector: 4, edge: 3 },
  { vector: 3, edge: 10 },
  { vector: 3, edge: 9 },
  { vector: 3, edge: 2 },
  { vector: 2, edge: 8 },
  { vector: 2, edge: 7 },
  { vector: 2, edge: 5 },
  { vector: 8, edge: 7 },
  { vector: 8, edge: 5 }, 
  { vector: 7, edge: 5 },
  { vector: 7, edge: 8 },
])

const stack = []
const dfs = (start_index) => {
  let index 
  graph[start_index].visited = true
  index = graph[start_index].head.edge
  stack.push(start_index)
  while(stack.length != 0) { 
    while(graph[index]) {
      let next = graph[index].head.edge
      if (next) {
        stack.push(index)
      }
      index = next
    }
    index = stack.pop()
    graph[index].head = graph[index].head.next
    if(graph[index].head == null) {
      graph[index] = null
    } 
  }
}
dfs(1)